lensless/benchmark.py

```diff
@@ -0,0 +1,270 @@
+        }
+    metrics_values = {key: 0.0 for key in metrics}
+    
+    # loop over batches
+    dataloader = DataLoader(dataset, batch_size=batchsize, pin_memory=(device != "cpu"))
+    model.reset()
+    for lensless, lensed in tqdm(dataloader):
+        lensless = lensless.to(device).squeeze()
+        lensed = lensed.to(device).permute(0, 3, 1, 2)
+
+        # compute predictions
+        with torch.no_grad():
+            if batchsize == 1:
+                model.set_data(lensless)
+                prediction = model.apply(plot=False, save=False, **kwargs)[None, :, :, :].permute(
+                    0, 3, 1, 2
+                )
+            else:
+                prediction = model.batch_call(lensless, **kwargs).permute(0, 3, 1, 2)
+
+        # normalization
+        prediction_max = torch.amax(prediction, dim=(1, 2, 3), keepdim=True)
+        prediction = prediction / prediction_max
+        lensed_max = torch.amax(lensed, dim=(1, 2, 3), keepdim=True)
+        lensed = lensed / lensed_max
+
+        # compute metrics
+        for metric in metrics:
+            metrics_values[metric] += metrics[metric](prediction, lensed).cpu().item()
+
+        model.reset()
+
+    # average metrics
+    for metric in metrics:
+        metrics_values[metric] /= len(dataloader)
+
+    return metrics_values
+
+
+if __name__ == "__main__":
+    from lensless import ADMM
+
+    downsample = 4
+    batchsize = 1
+    n_files = 10
+    image_ext = "npy"
+    n_iter = 100
+
+    # check if GPU is available
+    if torch.cuda.is_available():
+        device = "cuda"
+    else:
+        device = "cpu"
+
+    # download dataset if necessary
+    data = "data/DiffuserCam_Mirflickr_200_3011302021_11h43_seed11"
+    if not os.path.isdir(data):
+        print("No dataset found for benchmarking.")
+        try:
+            from torchvision.datasets.utils import download_and_extract_archive
+        except ImportError:
+            exit()
+        msg = "Do you want to download the sample dataset (725MB)?"
+
+        # default to yes if no input is given
+        valid = input("%s (Y/n) " % msg).lower() != "n"
+        if valid:
+            url = "https://drive.switch.ch/index.php/s/vmAZzryGI8U8rcE/download"
+            filename = "DiffuserCam_Mirflickr_200_3011302021_11h43_seed11.zip"
+            download_and_extract_archive(url, "data/", filename=filename, remove_finished=True)
+
+    # prepare dataset
+    psf_fp = os.path.join(data, "psf.tiff")
+    psf_float, background = load_psf(
+        psf_fp,
+        downsample=downsample,
+        return_float=True,
+        return_bg=True,
+        bg_pix=(0, 15),
+    )
+    dataset = ParallelDataset(
+        data, n_files=n_files, background=background, downsample=downsample, image_ext=image_ext
+    )
+
+    # prepare model
+    psf = torch.from_numpy(psf_float).to(device)
+    model = ADMM(psf, max_iter=n_iter)
+
+    # run benchmark
+    print(benchmark(model, dataset, batchsize=batchsize))
```

Comment:
@YohannPerron could you take a look and try? I tried to simplify `benchmark` API and what happens internally